### DISCUSSION TITLE
Database encryption implementation

### DISCUSSION THREAD

**Discussion by dAninAd1248** [uprm-inso4101-2023-2024-s2/semester-project-personalfinancemanager#159] [Ideas]
[Firebase already encrypts passwords used when signing in using password](https://firebaseopensource.com/projects/firebase/scrypt/), and encryption isn’t necessary when authentication through a third party service (Google and Facebook at the moment) is used. This means that we don’t have to encrypt passwords ourselves (although, when signing in through email and password is implemented, some parameters have to be set). However, the database should be encrypted to prevent anyone with access to the database to be able to see how broke our users are.

I propose we use asymmetric client-side encryption. In short, each user will have a public-private key pair. The public key is used to encrypt user data when writing to Firebase and will be stored in Firebase. The private key is used to decrypt user data retrieved from Firebase and will be stored in the user's device. A problem with this approach is that the user wouldn't be able to access their data from another device because this device doesn't have the private key. A good solution would be to send the private key from the old device to the new device. The private key would need to be encrypted when it's sent. A temporary password can be generated in the old device to do this and only this; it would be used to symmetrically encrypt the private key when sent from the old device and decrypt it from the new device. It will have to be manually entered by the user.

Specifically:
* When a new account is created:
    * Create a public private key pair.
    * Store the public key in Firebase and the private key in the user’s device (localStorage can be used here).
* When writing data to Firebase
    * Retrieve this user’s public key from Firebase.
    * Encrypt data.
    * Store encrypted data in Firebase.
* When reading data from Firebase
    * Retrieve encrypted data from Firebase.
    * Decrypt data using this user’s private key stored locally.
* When user logs in from a new device (this would be an entire feature)
    * Generate a temporary random password or passcode on old device.
    * Symmetrically encrypt private key using password as the key (directly, no key derivation is needed because this is being done manually).
    * Store encrypted private key in firebase
    * Retrieve encrypted private key on new device.
    * Show password to user on old device.
    * User enters temporary password on new device.
    * Use password to decrypt private key and store it locally on new device.
    * In order to check if password was entered correctly and successfully decrypted private key, also send a hash of the password using the same process, and check if the entered password’s hash matches the sent hash. If so, the password was correct.
    * Once private key is successfully stored (or user cancels private key transfer), delete the private key from Firebase (optional, it could also be kept there for backup). Also discard temporary password.

The following are only recommendations and are subject to change during development: Node.js’s crypto module can be used because the other good option is CryptoJS but it is discontinued. For asymmetric encryption (the data), RSA can be used; for symmetric encryption (private key transfer), AES-256 can be used; and for hashing, SHA-256 can be used.

Any questions, corrections, improvements, or suggestions are welcome. If you have a better idea for creating or transferring the private key, please leave a comment.

**Comment by JeanMontesSantiago** [comment]
Is this division of tasks appropriate for completing your feature?

### Task 1: Implement Key Pair Generation and Storage

1. **Generate Key Pair**: When a new account is created, use Node.js's crypto module to generate a public-private key pair. Ensure the keys are sufficiently strong for encryption purposes. For example, RSA keys should be at least 2048 bits long.
   
2. **Store Keys**: Store the public key in Firebase against the user's profile or account details. Store the private key securely on the user's device, considering localStorage for web applications. Ensure the storage method is secure and accessible only to the user's application.

### Task 2: Encrypt and Decrypt Data for Firebase

1. **Encrypt Data**: Before writing any user data to Firebase, retrieve the user's public key from Firebase. Use this key to encrypt the data client-side, ensuring that the data is unreadable by anyone without the corresponding private key.
   
2. **Decrypt Data**: When reading data from Firebase, decrypt it client-side using the private key stored on the user's device. Ensure the decryption process is secure and does not expose the private key.

### Task 3: Implement Secure Private Key Transfer Between Devices

1. **Generate Temporary Password**: On the old device, generate a temporary, random password or passcode. This password will be used to symmetrically encrypt the private key.

2. **Encrypt and Store Private Key**: Symmetrically encrypt the private key using the temporary password as the key and store the encrypted key in Firebase. Consider using AES-256 for strong encryption.

3. **Transfer and Decrypt Private Key**: On the new device, retrieve the encrypted private key from Firebase. Display the temporary password to the user on the old device and prompt them to enter it on the new device. Use the password to decrypt the private key and store it securely on the new device.

### Task 4: Ensure Security and Cleanup

1. **Verify Password**: To ensure the password was entered correctly, also send a hash of the password (using SHA-256) during the transfer process. Verify the entered password's hash matches the sent hash on the new device to confirm successful decryption.
   
2. **Cleanup**: Once the private key is successfully transferred (or the user cancels the transfer), consider deleting the encrypted private key from Firebase to ensure it does not become a security vulnerability. Also, discard the temporary password.

**Reply by dAninAd1248** [reply]
Task 1 and 2 look good, but it doesn't make much sense to separate the things in task 4 from the rest of the private key transfer process. I think it would be better if this process is divided between the client that sends the private key and the client that receives it, because these would be 2 separate react components. Something like

Task 3:
- Key transfer (sending) UI
- Private key encryption
- Password hashing
- Storing encrypted key and password hash in database
- Deleting stored key and hash if this screen is left

Task 4:
- Key transfer (receiving) and password entry UI
- Password verification using its hash
- Decrypting private key
- Storing private key

**Comment by diegoq48** [comment]
Whilst I agree that the private public key model is more elegant than just having the user directly send the password as the professor suggested I have a few concerns with this approach. We're relying on the user to secure their own device probably not going to happen. Also what happens if the user loses the key or the device or the password In what way could the keys be backed up without compromising the architechture of the program also how can we store this private key in a same way because correct me if im wrong the private key will just live in the filesystem unaltered why not say put it in a keyring. All of these are a bit difficult so worst case just fall back to the caveman approach from the prof also make sure to add proper key sizes for all the algorithms

**Reply by dAninAd1248** [reply]
So I was thinking about this and I realized that even if a malicious entity gains access to the private key, they can't do anything with it because they need the user’s data, and Firebase already takes care of making accessing the database impossible without authentication. So the purpose of encrypting the data is to protect the user data from anyone who has access, which apart from the user is only the developers and maybe some google admin (which I doubt). And I guess we can just trust that neither would try to hack a user’s device. So saving it in plaintext on the user's device would be safe. The only problem is, for example if localStorage is used, it could be cleared and the key would be lost. But I found a web API function that might prevent it from being cleared but in the worst case I could setup a database just for the key. And this is better than using a password to decrypt the data because this way the user doesn’t run the risk of forgetting their password and losing all their data. Tell me what you think.

EDIT: Just wanted to clarify some things.

["Firebase services encrypt data in transit using HTTPS and logically isolate customer data.", and it also encrypts Cloud Firestore (our database) at rest.](https://firebase.google.com/support/privacy#security_information), which means that, if we don't do something stupid like save the user's data in plaintext, the only people who have access are the user, developers, and "select employees who have a business purpose to access personal data".

About using a keyring, from my quick google search correct me if I'm wrong but the benefit of using one is for encrypting the key and using key rotation. Encrypting the key will require another key and we're back at square one, and key rotation might be overkill considering what I wrote above but it could be done, maybe as a small separate issue. If we're being paranoid about storing the key in plaintext [we could store it as a CryptoObject and store it in an IndexedDB database.](https://crypto.stackexchange.com/a/52488). Or store it in the device's keychain but it seems to me that's even more overkill.

The only major issue would be what if the user loses their device / uninstalls the web browser. For this, instead of using a temporary password generate in the moment, we can combine the private key approach with the user entered password approach. Still use the private key to decrypt data and still store it locally but upload it to the database as a backup and encrypt it using a password the user creates when making their account as the private key. Then, when the user signs in on a new device or web browser, download the backup private key and use the password to decrypt it. If the user forgets this password, they can reset it from a device that has the private key already stored by replacing the backup. If they forget their password AND lose the private key well nothing we can do there.

**Reply by dAninAd1248** [reply]
For RSA, keys with length of 2048 will be used as that is the standard. The rest of the lengths are in the original post (the number next to the algorithm).

**Comment by diegoq48** [comment]
Relying solely on Firebase's security features, while robust, overlooks a broader spectrum of potential attack vectors, including application vulnerabilities and unauthorized access to a user’s device. Saving encryption keys in plaintext on the device poses significant risks, as it leaves sensitive information vulnerable to extraction by attackers, undermining the protection offered by Firebase authentication. The use of localStorage for key storage is problematic due to its volatility, and while alternatives like IndexedDB and the device's keychain offer more security, they do not safeguard against all forms of device compromise. The consideration of keyrings and CryptoObjects introduces necessary complexity through additional layers of encryption and key management, but it neglects the essential practice of key rotation, which mitigates the risk of long-term key compromise. Furthermore, the strategy for handling lost devices and password resets, by combining a private key with a user-generated password for data recovery, strikes a balance between security and usability but requires meticulous implementation to prevent new vulnerabilities.
